### PR TITLE
[BEAM-8399] Add --hdfs_full_urls option

### PR DESCRIPTION
**Pull request by udim**
Introduces a new mode, enabled via `--hdfs_full_urls`.
In this mode, all hdfs:// URLs are expected to have a server designation in the first path element:
```
hdfs://server/path/...
or
hdfs://host:port/path/...
etc.
```

The server part is ignored; the connection config still comes from these flags:
```
--hdfs_host
--hdfs_port
--hdfs_user
```

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/io/hadoopfilesystem.py [hunk 1]
```diff
@@ -115,42 +116,59 @@ def __init__(self, pipeline_options):
       hdfs_host = hdfs_options.hdfs_host
       hdfs_port = hdfs_options.hdfs_port
       hdfs_user = hdfs_options.hdfs_user
+      self.full_urls = hdfs_options.hdfs_full_urls
```
Comment (zhitaoli) [inline sdks/python/apache_beam/io/hadoopfilesystem.py]: Make this private? `self._full_urls`

File: sdks/python/apache_beam/io/hadoopfilesystem_test.py [hunk 2]
```diff
@@ -323,7 +375,7 @@ def test_create_success(self):
     url = self.fs.join(self.tmpdir, 'new_file')
     handle = self.fs.create(url)
     self.assertIsNotNone(handle)
-    url = self.fs._parse_url(url)
+    _, url = self.fs._parse_url(url)
```
Comment (zhitaoli) [inline sdks/python/apache_beam/io/hadoopfilesystem_test.py]: Test `server` is None, here and below?

File: sdks/python/apache_beam/io/hadoopfilesystem_test.py [hunk 3]
```diff
@@ -579,6 +630,19 @@ def test_dict_options_missing(self):
...
+    pipeline_options = {
+        'hdfs_host': '',
+        'hdfs_port': 0,
+        'hdfs_user': '',
+        'hdfs_full_urls': 'invalid',
+    }
+
+    with
+    self.fs = hdfs.HadoopFileSystem(pipeline_options=pipeline_options)
+    self.assertFalse(self.fs.full_urls)
```
Comment (zhitaoli) [inline sdks/python/apache_beam/io/hadoopfilesystem_test.py]: Is this `with` a typo?

File: sdks/python/apache_beam/io/hadoopfilesystem.py [hunk 4]
```diff
@@ -163,19 +181,25 @@ def join(self, base_url, *paths):
...
   def _join(self, basepath, *paths):
     return posixpath.join(basepath, *paths)
 
   def split(self, url):
-    rel_path = self._parse_url(url)
+    server, rel_path = self._parse_url(url)
+    if server is None:
+      server = ''
+    else:
+      server = '/' + server
```
Comment (zhitaoli) [inline sdks/python/apache_beam/io/hadoopfilesystem.py]: Is this only expected to ever work in posix? If not, should we use `os.path.join`?

**Comment by zhitaoli** [comment]
ping? is it possible to get this proper merged?

File: sdks/python/apache_beam/io/hadoopfilesystem_test.py (hunk 2 above)
Comment (udim) [reply sdks/python/apache_beam/io/hadoopfilesystem_test.py]: There will be a separate `test_parse_url` to test these return values.

**Comment by udim** [comment]
R: @zhitaoli, @chamikaramj

File: sdks/python/apache_beam/io/hadoopfilesystem.py (hunk 4 above)
Comment (zhitaoli) [reply sdks/python/apache_beam/io/hadoopfilesystem.py]: Nevermind. I think if we consider hfds as a constraint this is fine.

**Comment by udim** [comment]
Run Portable_Python PreCommit

**Comment by udim** [comment]
Run PythonLint PreCommit

**Comment by udim** [comment]
Run PythonFormatter PreCommit

**Comment by udim** [comment]
Run PythonFormatter PreCommit

**Comment by udim** [comment]
Run Python PreCommit

**Comment by udim** [comment]
Run Python PreCommit

**Comment by chamikaramj** [comment]
Run Python PreCommit

**Comment by udim** [comment]
Run Python PreCommit

**Pull request merged**